src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -209,30 +207,23 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
     else if (that.isEmpty) this.asInstanceOf[That]
     else that match {
       case thatHash: HashMap[A, B] =>
-        val merger: Merger[A, B] = HashMap.liftMerger[A, B](null)
-        // merger prefers to keep than replace
-        // so we invert
-        (this.merge0(thatHash, 0, merger.invert)).asInstanceOf[That]
+        (this.merge0(thatHash, 0, HashMap.concatMerger.invert.asInstanceOf[Merger[A, B]])).asInstanceOf[That]
 
       case that:HasForeachEntry[A, B] =>
         object adder extends Function2[A, B, Unit] {
-          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
-          val merger = HashMap.liftMerger[A, B](null)
-
+          var result: HashMap[A, B] = HashMap.this.asInstanceOf[HashMap[A, B]]
           override def apply(key: A, value: B): Unit = {
-            result = result.updated0(key, computeHash(key), 0, value, null, merger)
+            result = result.updated0(key, computeHash(key), 0, value, null, HashMap.concatMerger.invert.asInstanceOf[Merger[A, B]])
           }
         }
         that foreachEntry adder
         adder.result.asInstanceOf[That]
       case that =>
         object adder extends Function1[(A,B), Unit] {
-          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
-          val merger = HashMap.liftMerger[A, B](null)
-
+          var result: HashMap[A, B] = HashMap.this.asInstanceOf[HashMap[A, B]]
           override def apply(kv: (A, B)): Unit = {
             val key = kv._1
-            result = result.updated0(key, computeHash(key), 0, kv._2, kv, merger)
+            result = result.updated0(key, computeHash(key), 0, kv._2, kv, HashMap.concatMerger.invert.asInstanceOf[Merger[A, B]])
```

Comment:
can we restore `merger` as a field to avoid the method calls

could be a post-merge followup though as it doesnt affect functionality, and it would give time to check the performance